### ISSUE TITLE
Column access rules don’t apply to Python formulas

### ISSUE THREAD

**Issue by jperon** [gristlabs/grist-core#305]
I don’t know wether it’s a bug or a feature, but when a Python formula (either in another column, or in a conditional style) referes to a "CENSORED" column, it can access its content. This document is a minimal example:
https://docs.getgrist.com/gPsGvhQYnzcG/ACL

Table1 has a "Private" column, whose access is restricted to my user, and it doesn’t show (as expected); but the conditional row style and the "Reveal_private" column have access to its value.

It could be a feature (allowing some formulas while hiding the source), but as it may leak sensitive data, I thought it was worth mentioning.

**Comment by dsagal** [comment]
**TL;DR** Deny the "Structure" permission (`S`) to anyone with limited access to the document, to block their ability to create or change formulas. See https://support.getgrist.com/access-rules/#lock-down-structure.

---

Indeed, Grist formulas are not limited by access rules. Formulas are part of the single document content -- they don't evaluate differently for different users. For this reason, formulas are given full access to the document. What you can limit is someone's access to formula results.

In part, this is a feature: e.g. you may have columns like "Balance" and "Amount Paid", which some users don't have access to, but a calculated column "Is Fully Paid" (computed from the two protected columns) can be accessible to all.

The biggest danger of leaks is from the ability to write formulas, since that indeed enables one to gain access to _any_ data in the document. That's why the section [Lock down structure](https://support.getgrist.com/access-rules/#lock-down-structure) is the first rule we recommend adding for a document with access rules. See the "Important" note at the end of that section, which explains this same point about formulas.

In the future, we'd like to make it easier to lock down the ability to write formulas, and harder to miss accidentally (admittedly, currently it's too easy to miss without reading documentation attentively).

**Comment by jperon** [comment]
I had seen the lock-down indication.
In fact, what I found most "fragile" was the conditional style point: if an owner forgets that point and defines a row style instead of a cell style, there’s some room for unintended leaks.
